### DISCUSSION TITLE
HV programmers

### DISCUSSION THREAD

**Discussion by iansexton** [SpenceKonde/megaTinyCore#349] [Q&A]
Anyone using a HV programmer?
I bought Dlloyds offering & it doesn't work for me :(

**Comment by mechatroniks-git** [comment]
I'm thinking about getting the LeoNerds.

**Reply by iansexton** [reply]
Let us know if/when you get it.
It's very tempting - he's in the UK & I am too so the postage costs are reasonable for a change!!

**Reply by mechatroniks-git** [reply]
Will do. Not sure if I need it, I tried to test setting the lockbit to something other than 0xC5 on my ATtiny404, adding this to the program command. I think it works, but can't seem to erase the chip. Am I right that I need the HV programmer to erase?

.......avrdude.conf -v -pattiny404 -cjtag2updi -PCOM1 -Ufuse2:w:0x02:m -Ufuse6:w:0x04:m -Ufuse8:w:0x00:m -**Ulock:w:0x5C:m** -Uflash:w:C:\Users.......

**Reply by jvasileff** [reply]
The spec says:

> Important:  The only way to unlock a device is through a CHIPERASE. No application data are retained.

So it seems like `avrdude -e ...` should be sufficient. Is that what you tried?

**Reply by mechatroniks-git** [reply]
This is what I did, didn't seem to work. 

```
C:\Users\m\AppData\Local\Arduino15\packages\arduino\tools\avrdude\6.3.0-arduino18/bin/avrdude -C C:\Users\m\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.2.9/avrdude.conf -v  -p t404 -c jtag2updi -P COM1  -e

avrdude.exe: Version 6.3-20201216
             Copyright (c) 2000-2005 Brian Dean, http://www.bdmicro.com/
             Copyright (c) 2007-2014 Joerg Wunsch

             System wide configuration file is "C:\Users\pc\AppData\Local\Arduino15\packages\megaTinyCore\hardware\megaavr\2.2.9/avrdude.conf"

             Using Port                    : COM1
             Using Programmer              : jtag2updi
JTAG ICE mkII sign-on message:
Communications protocol version: 1
M_MCU:
  boot-loader FW version:        1
  firmware version:              6.00
  hardware version:              1
S_MCU:
  boot-loader FW version:        1
  firmware version:              6.00
  hardware version:              1
Serial number:                   00:00:00:00:00:00
Device ID:                       JTAGICE mkII
             AVR Part                      : ATtiny404
             Chip Erase delay              : 0 us
             PAGEL                         : P00
             BS2                           : P00
             RESET disposition             : dedicated
             RETRY pulse                   : SCK
             serial program mode           : yes
             parallel program mode         : yes
             Timeout                       : 0
             StabDelay                     : 0
             CmdexeDelay                   : 0
             SyncLoops                     : 0
             ByteDelay                     : 0
             PollIndex                     : 0
             PollValue                     : 0x00
             Memory Detail                 :

                                      Block Poll               Page                       Polled
               Memory Type Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
               ----------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
               signature      0     0     0    0 no          3    0      0     0     0 0x00 0x00
               prodsig        0     0     0    0 no         61   61      0     0     0 0x00 0x00
               fuses          0     0     0    0 no          9   10      0     0     0 0x00 0x00
               fuse0          0     0     0    0 no          1    0      0     0     0 0x00 0x00
               fuse1          0     0     0    0 no          1    0      0     0     0 0x00 0x00
               fuse2          0     0     0    0 no          1    0      0     0     0 0x00 0x00
               fuse4          0     0     0    0 no          1    0      0     0     0 0x00 0x00
               fuse5          0     0     0    0 no          1    0      0     0     0 0x00 0x00
               fuse6          0     0     0    0 no          1    0      0     0     0 0x00 0x00
               fuse7          0     0     0    0 no          1    0      0     0     0 0x00 0x00
               fuse8          0     0     0    0 no          1    0      0     0     0 0x00 0x00
               lock           0     0     0    0 no          1    0      0     0     0 0x00 0x00
               data           0     0     0    0 no          0    0      0     0     0 0x00 0x00
               usersig        0     0     0    0 no         32   32      0     0     0 0x00 0x00
               flash          0     0     0    0 no       4096   64      0     0     0 0x00 0x00
               eeprom         0     0     0    0 no        128   32      0     0     0 0x00 0x00

             Programmer Type : JTAGMKII_PDI
             Description     : JTAGv2 to UPDI bridge
             M_MCU hardware version: 1
             M_MCU firmware version: 6.00
             S_MCU hardware version: 1
             S_MCU firmware version: 6.00
             Serial number:          00:00:00:00:00:00
             Vtarget         : 5.0 V

avrdude.exe: jtagmkII_initialize(): Cannot locate "flash" and "boot" memories in description
avrdude.exe: jtagmkII_reset(): timeout/error communicating with programmer (status -1)
avrdude.exe: initialization failed, rc=-1
             Double check connections and try again, or use -F to override
             this check.

avrdude.exe: jtagmkII_close(): timeout/error communicating with programmer (status -1)
avrdude.exe: jtagmkII_close(): timeout/error communicating with programmer (status -1)

avrdude.exe done.  Thank you.
```

**Reply by jvasileff** [reply]
I gave it a try with an ATtiny 814, but I seem to have bricked (for now) my chip, but in a different way than yours. So maybe in your case there is something more going on than just the lock.

I locked mine with:

```
% avrdude \
    -C./avrdude.conf \
    -v -pattiny814 -cjtag2updi -P/dev/cu.SLAB_USBtoUART \
    -U lock:w:0x99:m

...

avrdude: verifying ...
avrdude: 1 bytes of lock verified

avrdude done.  Thank you.
```

and then attempted to unlock with:

```
% avrdude \
    -C./avrdude.conf \
    -v -pattiny814 -cjtag2updi -P/dev/cu.SLAB_USBtoUART \
    -e

...

avrdude: AVR device initialized and ready to accept instructions

Reading |                                                    | 0% 0.00savrdude: jtagmkII_program_enable(): bad response to enter progmode command: RSP_ILLEGAL_MCU_STATE
avrdude: jtagmkII_program_enable(): bad response to enter progmode command: RSP_ILLEGAL_MCU_STATE
avrdude: jtagmkII_read_byte(): bad response to read memory command: RSP_ILLEGAL_MCU_STATE
avr_read(): error reading address 0x0000
    read operation not supported for memory "signature"
avrdude: error reading signature data for part "ATtiny814", rc=-2
avrdude: error reading signature data, rc=-1
avrdude: jtagmkII_program_disable(): bad response to leave progmode command: RSP_ILLEGAL_MCU_STATE

avrdude done.  Thank you.
```

It's worth noting that the jtag2updi [README](https://github.com/ElTangas/jtag2updi/blob/master/README.md) claims this should work:

```
avrdude -c jtag2updi -P com7 -p t1614 -e
```

**Reply by mechatroniks-git** [reply]
Looks like we did the same thing. My device is still running it's code, just can't reprogram or erase it. So I think the lockbit is doing the job, the ERASE is failing. Perhaps I'll submit an issue over at jtag2updi. 

If it is bricked, will the HV UPDI rescue it? I'm still unsure quite what the HV programmer can do.

**Reply by jvasileff** [reply]
Well, yours fails right away:

```
avrdude.exe: jtagmkII_reset(): timeout/error communicating with programmer (status -1)
avrdude.exe: initialization failed, rc=-1
...
```

while mine connects, then fails:

```
avrdude: AVR device initialized and ready to accept instructions

Reading |                                                    | 0% 0.00savrdude: jtagmkII_program_enable(): bad response to enter progmode command: RSP_ILLEGAL_MCU_STATE
avrdude: jtagmkII_program_enable(): bad response to enter progmode command: RSP_ILLEGAL_MCU_STATE
...
```

I'm not an expert on this, but I think the HV is only needed if you change the default for the UDPI pin to RESET or GPIO (fuse 5, reset pin configuration). But who knows, maybe it can also wake up a misbehaving chip?

**Comment by SpenceKonde** [comment]
Uhhhhh... guys...  hv programming has nothing to do with lock and unlock....

To unlock a chip you have locked, you shoud be able to just do a chip erase
cycle... normal updi, it will erase the flash, then reset the lock it's to
unlocked.

Hv updi is for reprogramming after configuring the updi pin to act as reset
or gpio...
____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Reply by mechatroniks-git** [reply]
Copy that. Thanks for the clarification. It appears that -e erase no longer works with AVRDUDE on this device here. Although, my first attempt to write to the lockbit was using Ulock:w:0xC5:m, which I figured was safe to do, as that is a valid key. Then I did 0x5C, which seemed to lock it. Now erase is not working. 

Before I go (potentially) bricking another device, what else can I try?

**Reply by jvasileff** [reply]
Although our problems may be different, I was able to un-brick mine with `pyupdi` after setting the baud rate to 57600 (it didn't work in my first attempt at a different rate, but it could be a coincidence that setting the baud rate to 57600 fixed the problem).

The command I used:

```
pyupdi \
        -d attiny814 \
        -c /dev/cu.SLAB_USBtoUART4 \
        -b 57600 \
        -e -v
```

**Reply by mechatroniks-git** [reply]
This works with the LeoNerd HV UPDI device (in non-HV mode). Thanks for the suggestion. It may have also taken 2 tries. 

Any idea how to read/write the EEPROM & USERROW with pyupdi? I see how to write, but would like to read, so I can get a file format that works & be able to verify.

`  -fe EEPROM, --eeprom EEPROM
                        Intel HEX file to write to EEPROM.`

**Comment by SpenceKonde** [comment]
I have a feeling that trying to burn bootloader with the pyupdi style programmer option would work, too.. Also try with an official microchip programmer, rather than the jtag2updi. Thinking about the communication between avrdude and jtag2updi, I'm not sure there's a way for it to get to the chip erase command when it determines the chip is locked >.>

**Reply by jvasileff** [reply]
To share results from additional testing:

"Burn Bootloader" didn't work for me (error below). I also tried invoking`prog.py` at the command line without the `--fuses` option, and that failed in the same way.

I've found `pyupdi` to fail more often than it works when erasing a locked chips, with the success rate seeming to increase with lower baud rates, like 9600. When the device is unlocked, `pyupdi` works for me 100% of the time.

I read somewhere (I think in a closed jtag2updi GitHub issue) that `avrdude` can be used to erase locked chips using terminal mode, but I didn't test that (terminal mode requires compiling jtag2updi with `-DDISABLE_HOST_TIMEOUT`, which I don't want to do).

```
/Users/jvasileff/Library/Arduino15/packages/megaTinyCore/tools/python3/3.7.2-post1/python3 -u /Users/jvasileff/Library/Arduino15/packages/megaTinyCore/hardware/megaavr/2.2.9/tools/prog.py -t uart -u /dev/cu.SLAB_USBtoUART -d attiny814 -v --fuses 0:0x00 1:0x00 2:0x01 4:0x00 5:0b11000101 6:0x04 7:0x00 8:0x00 -aerase
pymcuprog.programmer - INFO - Setting up programming session for 'attiny814'
pymcuprog.deviceinfo.deviceinfo - INFO - Looking for device attiny814
pymcuprog.serialupdi.physical - INFO - Opening port '/dev/cu.SLAB_USBtoUART' at '115200' baud
...
pymcuprog.serialupdi.application - ERROR - Timeout waiting for device to unlock
pymcuprog.nvm - ERROR - Device is locked.
Error:
Failed to enter NVM programming mode: device is locked
pymcuprog.nvm - INFO - No specific initializer for this provider
Pinging device...
pymcuprog.programmer - INFO - Reading device ID...
...
pymcuprog.pymcuprog_errors.PymcuprogError: Error with st_ptr
Error while burning bootloader.
```

**Reply by mechatroniks-git** [reply]
Jusy confirmed terminal mode erase worked on locked chip with jtag2updi. 9600 and 38.4k baud did not work on command line, even with -F, it didn't like the 24bit chip ID not matching.

**Comment by SpenceKonde** [comment]
There's a new version of the pymcuprog tool and arduino bridge that shoulod be able to erase locked chips (in master branch). Totally untested!!

**Reply by mechatroniks-git** [reply]
This worked, which I downloaded from here: https://pypi.org/project/pymcuprog/

`pymcuprog erase -d attiny404 -t uart -u COM8`

pymcuprog seems way more robust and full-featured than jtag2updi or pyupdi or avrdude. 

Does the version you have here do it any differently? If so, is there a PIP or instructins to install?

**Reply by SpenceKonde** [reply]
Yes, that's WHY we used it instead of one of those other tools!

Ours is a stock version of pymcuprog with, IIRC, only two changes - it doesn't assert DTR and RTS (almost everything does; but because we have full control over pymcuprog, we can have it not do that, so that we could wire one of those lines to a common mux IC and have it connect to the resistor/diode when RTS is not asserted, and the pins of a normal serial adapter when it is), and there was one point where, with verbose output enabled, and it's forwarding some of the data it reads out of the chip, it would sent an unescaped null (character 0), which broke almost application one tried to pipe the output to, since null is used as the string terminator.

It is distributed only as part of these cores, it's Microchip's tool, and while they give it away for free, it's not open source. We make minimal tweaks to it as needed to make it work here, but it's certainly not our place to go making an independent version.

The work we have done is mostly the bridge that goes between Arduino IDE and pymcuprog... But yeah, the version currently in the repo's master branch, which I hope to be able to release tonight, should automatically unlock and erase locked chips. I haven't tested it and have wayyyyyy too many action items on my list to take that on.

**Reply by mechatroniks-git** [reply]
I spoke too soon. I cannot get the following commands to work.

1. Chip erase when locked doesn't work. `pymcuprog -t uart -u COM8 -d attiny404 erase -C`. 
2. When the chip is locked, can't update the userrow: `pymcuprog -t uart -u COM8 -d attiny404 write -U -f userrow2.hex` fails. 
3. Can't figure out how to lock chip with pymcuprog

I can use pyupdi to lock and erase the chip.
